bin/find-fixable-orphans.py

```diff
@@ -0,0 +1,293 @@
+    this_s3_bucket_raw_keys_list = [
+        s3_object.key for s3_object in s3_resource_bucket.objects.all()
+    ]
+
+    print(
+        f"{len(this_s3_bucket_raw_keys_list)} rows retrieved",
+        file=sys.stderr,
+    )
+
+    return this_s3_bucket_raw_keys_list
+
+
+# Get the keys present and absent across NZSL Signbank and S3, to dictionary
+def create_all_keys_dict(this_nzsl_raw_keys_dict, this_s3_bucket_raw_keys_list):
+    print(
+        "Getting keys present and absent across NZSL Signbank and S3 ...",
+        file=sys.stderr,
+    )
+    this_all_keys_dict = {}
+
+    # Find S3 keys that are present in NZSL, or absent
+    for video_key in this_s3_bucket_raw_keys_list:
+        dict_row = this_nzsl_raw_keys_dict.get(video_key, None)
+        if dict_row:
+            # NZSL glossvideo record for this S3 key
+            this_all_keys_dict[video_key] = [
+                True,  # NZSL PRESENT
+                True,  # S3 PRESENT
+            ] + dict_row
+        else:
+            # S3 key with no corresponding NZSL glossvideo record
+            this_all_keys_dict[video_key] = [
+                False,  # NZSL Absent
+                True,  # S3 PRESENT
+            ] + [""] * 6
+
+    # Find NZSL keys that are absent from S3 (present in both handled above)
+    for video_key, dict_row in this_nzsl_raw_keys_dict.items():
+        if video_key not in this_s3_bucket_raw_keys_list:
+            # gloss/glossvideo record with no corresponding S3 key
+            # Either:
+            # video_key is real, but the S3 object is missing
+            # video_key is fake (to handle the FULL JOIN) and this gloss/glossvideo never had an S3 object
+            this_all_keys_dict[video_key] = [
+                True,  # NZSL PRESENT
+                False,  # S3 Absent
+            ] + dict_row
+
+    return this_all_keys_dict
+
+
+def find_orphans():
+    all_keys_dict = create_all_keys_dict(
+        get_nzsl_raw_keys_dict(), get_s3_bucket_raw_keys_list()
+    )
+    print("Finding fixable orphans", file=sys.stderr)
+
+    print(CSV_DELIMITER.join(GLOBAL_COLUMN_HEADINGS))
+
+    # Traverse all the NZSL Signbank glosses that are missing S3 objects
+    for video_key, [
+        key_in_nzsl,
+        key_in_s3,
+        gloss_idgloss,
+        gloss_created_at,
+        gloss_id,
+        video_id,
+        gloss_public,
+        video_public,
+    ] in all_keys_dict.items():
+
+        if not key_in_nzsl:
+            # This is an S3 object, not a Signbank record
+            continue
+
+        if key_in_s3:
+            # This Signbank record already has an S3 object, all is well
+            continue
+
+        # Business rule
+        if int(gloss_id) < 8000:
+            continue
+
+        # The gloss_id is the only reliable retrieval key at the Signbank end
+        gloss = Gloss.objects.get(id=gloss_id)
+        gloss_name = gloss.idgloss.split(":")[0].strip()
+        video_path = gloss.get_video_path()
```

Comment:
<!-- Suggestions propose improvements to the current subject. It's important to be explicit and clear on what is being suggested and why it is an improvement. Consider using patches and the blocking or non-blocking decorations to further communicate your intent.  -->
**suggestion:** Don't get the video path

The `get_video_path` does one or 2 additional DB lookups that return substantial amount of data. You only really need to know if the video path exists. So do that instead:

```python
video_path_exists = gloss.glossvideo_set.exists()
```
‏

Reply:
Great idea, done!